### ISSUE TITLE
what's the role of "mask"  in train and test.

### ISSUE THREAD

**Issue by wadesunyang** [DandilionLau/Selfie-Friendly-Abstraction#4]
Hello,        
         I read your paper and test some image,  the result performs very good.   And I have a question when i prepare train my own model.

         What the effect of the "mask",  since the paper didn't mention about it, and i'm a little confuse about it.

         Another question is the model showed in paper is trained only by the 40 images?

        Thank you very much, wish to your replay.

**Comment by DandilionLau** [comment]
Hi,
1. When our code was released, there was no implementation of 'subpixel upsampling layer' in caffe, so we use the nearest neighbor upsampling layer with a selective mask to do the job of subpixel upsampling.
2. Yes. Our training set only contains 40 selfie photos. For each style, we have 40 reference style images generated from Prisma. I am not such how much performance gain will you have with a larger training set, but I think it is worth trying.

**Comment by wadesunyang** [comment]
Thanks for your reply.

And when i train the model use train_reshape.m,  it has 4000K iter  (pass=1:100, file = 1:40, batch=1:1000) , which is different from what the paper says(100K iterations with lr 10-8 and 10-9), the most i want to know is how to set the learning rate , because it didn't match with the paper says.  

thank you .

**Comment by DandilionLau** [comment]
The lr and max_iter is defined in the solver file of the training dataset. The 40 files are actually a split of one giant training file (due to RAM limitation, we use 40 smaller files to split). Although it may seem like 4000K iterations, the actual number of iterations is 100K.

**Comment by wadesunyang** [comment]
I know what you means, but it didn't stop since the “for  loop” is still run after 100K, although I didn't test the parameters of the network is still change or not. I will manual stop it when the iterations is 100K.
And why the lr is so small , because in my impression, the lr usually starts from 10e-3 or 10e-4 when do the task of detection or classification.  I have few experience in the area of Abstraction, could you tell me the reason why the lr set so small ?
sorry to bother you, thank you very much.

**Comment by DandilionLau** [comment]
I am not sure in our previous experiments (with an older version of caffe) the 'didn't stop' problem emerged or not... But since the model size is very tiny (with little time per iteration), I guess the safest choice is to wait for the script to terminate. Of course, you can manually test it on 100K and see the results...

We choose this lr from many times of experiments. I believe you can increase it a little bit but the model tends to collapse for lr of 1e-3 or 1e-4. One possible reason is that the gradient domain is different from the RGB domain, with less stable high-dimensional geometry for optimization (it is like optimizing the second-order derivative of an image).

**Comment by wadesunyang** [comment]
Thank you .   I trained the network with your guide, and get good performance. 
And to reduce the computation, I use the test image of 256*256, but the result is not good as 512*512,  So if i want to compute a video in real time , i need to change a small network and maybe use the v and h only in gray channel, any advice for the small net ?
I think it needn't use so much channel like 512, since in my understanding, you use 512 is to upsample with mask,  if i use pytorch which has the upsample layer, i can use only 32 channels before upsample, then upsample to same 32 channels,  it can save lots of computation，is this understanding right ?

**Comment by DandilionLau** [comment]
Sure. Initializing the mask matrix taking a fairly large amount of time in Matlab. Our paper is just a poof-of-concept and there are potentially many things to do for accelerating the model... I believe using the official subpixel layer would be much faster (I guess the performance would be close to 'real-time' for images of 512*512 in PyTorch).
